cdap-explore/src/main/java/co/cask/cdap/explore/service/ExploreTableService.java

```diff
@@ -0,0 +1,449 @@
+      "ALTER TABLE %s ADD PARTITION %s LOCATION '%s'",
+      getDatasetTableName(datasetID), generateHivePartitionKey(partitionKey), fsPath);
+
+    LOG.debug("Add partition for key {} dataset {} - {}", partitionKey, datasetID, addPartitionStatement);
+
+    return exploreService.execute(datasetID.getNamespace(), addPartitionStatement);
+  }
+
+
+  /**
+   * Adds multiple partitions to the Hive table for the given dataset.
+   *
+   * @param datasetID the ID of the dataset to add partitions to
+   * @param partitions a map of partition key to partition path
+   * @return the query handle for adding partitions to the dataset
+   * @throws ExploreException if there was an exception adding the partition
+   * @throws SQLException if there was a problem with the add partition statement
+   */
+  public QueryHandle addPartitions(Id.DatasetInstance datasetID,
+                                   Map<PartitionKey, String> partitions) throws ExploreException, SQLException {
+    if (partitions.isEmpty()) {
+      return QueryHandle.NO_OP;
+    }
+    StringBuilder statement = new StringBuilder()
+      .append("ALTER TABLE ")
+      .append(getDatasetTableName(datasetID))
+      .append(" ADD");
+    for (Map.Entry<PartitionKey, String> partitionEntry : partitions.entrySet()) {
+      statement.append(" PARTITION")
+        .append(generateHivePartitionKey(partitionEntry.getKey()))
+        .append(" LOCATION '")
+        .append(partitionEntry.getValue())
+        .append("'");
+    }
+
+    LOG.debug("Adding partitions for dataset {}", datasetID);
+
+    return exploreService.execute(datasetID.getNamespace(), statement.toString());
+  }
+
+  /**
+   * Drop a partition from the Hive table for the given dataset.
+   *
+   * @param datasetID the ID of the dataset to drop the partition from
+   * @param partitionKey the partition key to drop
+   * @return the query handle for dropping the partition from the dataset
+   * @throws ExploreException if there was an exception dropping the partition
+   * @throws SQLException if there was a problem with the drop partition statement
+   */
+  public QueryHandle dropPartition(Id.DatasetInstance datasetID, PartitionKey partitionKey)
+    throws ExploreException, SQLException {
+
+    String dropPartitionStatement = String.format(
+      "ALTER TABLE %s DROP PARTITION %s",
+      getDatasetTableName(datasetID), generateHivePartitionKey(partitionKey));
+
+    LOG.debug("Drop partition for key {} dataset {} - {}", partitionKey, datasetID, dropPartitionStatement);
+
+    return exploreService.execute(datasetID.getNamespace(), dropPartitionStatement);
+  }
+
+  private String getStreamTableName(Id.Stream streamId) {
+    return getHiveTableName(String.format("stream_%s", streamId.getName()));
+  }
+
+  private String getDatasetTableName(Id.DatasetInstance datasetID) {
+    return getHiveTableName(String.format("dataset_%s", datasetID.getId()));
+  }
+
+  private String getHiveTableName(String name) {
```

Comment:
This seems to be used directly only for datasets with ``ObjectMappedTableModule`` . Don't we have a ``Id.DatasetInstance`` for them?